subsystem/src/lib.rs

```diff
@@ -0,0 +1,316 @@
+    ///         };
+    ///     }
+    /// });
+    /// # let _ = subsystem.call(|()| ());  // Fix the call type to avoid amnbiguity.
+    /// ```
+    pub fn start<T: 'static + Send, F: 'static + Send + Future<Output = ()>>(
+        self,
+        subsystem: impl 'static + Send + FnOnce(CallRequest<T>, ShutdownRequest) -> F,
+    ) -> Subsystem<T> {
+        // Name strings
+        let manager_name = self.manager.name;
+        let subsys_name = self.subsystem_name;
+        // Shutdown-related channels
+        let shutting_down_tx = self.manager.shutting_down_tx.clone();
+        let shutdown_rq = ShutdownRequest(self.manager.shutdown_request_tx.subscribe());
+        // Call related channels
+        let (action_tx, action_rx) = mpsc::channel(self.call_queue_capacity);
+        let call_rq = CallRequest(action_rx);
+
+        task::spawn(async move {
+            log::info!("Subsystem {}/{} started", manager_name, subsys_name);
+
+            // Perform the subsystem task.
+            subsystem(call_rq, shutdown_rq).await;
+
+            // Signal the intent to shut down to the other parts of the application.
+            let _ = shutting_down_tx.send(()).await;
+
+            // Close the channel to signal the completion of the shutdown.
+            std::mem::drop(shutting_down_tx);
+
+            log::info!("Subsystem {}/{} terminated", manager_name, subsys_name);
+        });
+
+        Subsystem::new(action_tx)
+    }
+
+    /// Start a passive subsystem.
+    ///
+    /// A passive subsystem does not interact with the environment on its own. It only serves calls
+    /// from other subsystems.
+    pub fn start_passive<T: 'static + Send>(self, mut obj: T) -> Subsystem<T> {
+        self.start(|mut call_rq, mut shutdown_rq| async move {
+            loop {
+                tokio::select! {
+                    () = shutdown_rq.recv() => { break; }
+                    call = call_rq.recv() => { call(&mut obj); }
+                }
+            }
+        })
+    }
+}
+
+/// Internal action type sent in the channel.
+type Action<T> = Box<dyn FnOnce(&mut T) + Send>;
+
+/// Call request
+pub struct CallRequest<T>(mpsc::Receiver<Action<T>>);
+
+impl<T: 'static + Send> CallRequest<T> {
+    /// Receive an external call to this subsystem.
+    pub async fn recv(&mut self) -> Action<T> {
+        match self.0.recv().await {
+            // We have a call, return it
+            Some(action) => action,
+            // All handles to this subsystem dropped, suspend call handling.
+            None => std::future::pending().await,
+        }
+    }
+}
+
+/// Shutdown request
+pub struct ShutdownRequest(broadcast::Receiver<()>);
+
+impl ShutdownRequest {
+    /// Receive a shutdown request.
+    pub async fn recv(&mut self) {
+        match self.0.recv().await {
+            Err(broadcast::error::RecvError::Lagged(_)) => {
+                panic!("Multiple shutdown broadcast requests issued")
```

Comment:
Why is this a panic? Is it a logic error for several subsystems to independently signal shutdown? In general, what will make a subsystem want to shut down?